### DISCUSSION TITLE
graph_traversal documentation

### DISCUSSION THREAD

**Discussion by kimdn** [zhonge/cryodrgn#83] [Q&A]
I ran as 
`cryodrgn graph_traversal data vae128_z8_e400_beta_4_amp`
but it says

```
usage: cryodrgn graph_traversal [-h] --anchors ANCHORS [ANCHORS ...]
                                [--max-neighbors MAX_NEIGHBORS]
                                [--avg-neighbors AVG_NEIGHBORS]
                                [--batch-size BATCH_SIZE]
                                [--max-images MAX_IMAGES] -o PATH.TXT --out-z
                                Z.PATH.TXT
                                data
cryodrgn graph_traversal: error: the following arguments are required: --anchors, -o, --out-z

```
Can I know what is anchors?
Thank you

**Comment by zhonge** [comment]
`--anchors` are indices of the latent embeddings that will be used as anchors in the graph traversal algorithm.

Here's an example use case, where the kmeans cluster centers are used as the anchor points:

```
$ cd analyze.[epoch] # assuming you've already run cryodrgn analyze
$ cryodrgn graph_traversal ../z.[epoch].pkl --anchors $(cat kmeans20/centers_ind.txt) -o graph_traversal/path.txt --out-z graph_traversal/z.path.txt
```

Note that I sometimes reorder the cluster centers before running the graph traversal algorithm, so that the generated trajectory is shorter.

Also note that you could specify arbitrary data points as the anchor points, e.g. you could manually find the indices of desired data points from the jupyter notebook (in the interactive visualization sections) that you want as the end points of the trajectory.

The algorithm builds a nearest neighbor graph between all the latent embeddings, and then performs dijkstra's algorithm to find the shortest path on the graph between the anchors nodes. The idea is to find a path between the anchors points in latent space while remaining on the data manifold, since we don't want to generate structures from empty regions of the latent space.

Once you've run `cryodrgn graph_traversal`, you can generate `.mrc` files with `cryodrgn eval_vol`, e.g.

```
$ cd graph_traversal
$ cryodrgn eval_vol ../../weights.[epoch].pkl -c ../../config.pkl --zfile z.path.txt -o .
```
